Fixed #31007 -- Allowed specifying type of auto-created primary keys.

orf opened the pull request:
Ticket: https://code.djangoproject.com/ticket/31007

To lessen the impact of this change with Django contrib modules, I have set `default_model_pk` to `AutoField` in all the AppConfigs, and `UUIDField` will add a `uuid4` default callable if it is auto-created.

auvipy commented:
I quite like the approach

caioariede commented on django/apps/config.py:
```diff
@@ -214,3 +214,8 @@ def ready(self):
         Override this method in subclasses to run code when Django starts.
         """
+
+    @property
+    def model_default_pk(self):
+        from django.conf import settings
+        return settings.MODEL_DEFAULT_PK
```
I feel like this setting is out of context here. Should `AppConfig` only provide information in the context of an app?

If I want to find out what kind of primary key a given app is asking for, I could look for `AppConfig.model_default_pk`, but if we're overriding if with a project setting it can cause some confusion.

orf replied:
The _default_ model PK field is defined in `settings.MODEL_DEFAULT_PK`, which applies to all applications thanks to this property. This can be overridden on a per-app basis if they ship migrations. Without this if an app ships migrations then changing the `MODEL_DEFAULT_PK` will cause new migrations to be generated, which isn't appropriate for third party apps.

caioariede commented on django/db/models/options.py:
```diff
@@ -250,7 +251,9 @@ def _prepare(self, model):
                 self.setup_pk(field)
             else:
-                auto = AutoField(verbose_name='ID', primary_key=True, auto_created=True)
+                pk_setting = getattr(self.app_config, "model_default_pk", settings.MODEL_DEFAULT_PK)
+                pk_class = import_string(pk_setting)
+                auto = pk_class(verbose_name='ID', primary_key=True, auto_created=True)
```
🐛 I think there's a soft bug here. Right now `model_default_pk` is a property that already defaults to `settings.MODEL_DEFAULT_PK`. In this case the third argument in the `getattr(...)` call will never be used right?

orf replied:
There is one exclusion to this: the `Migration` model itself. `self.app_config` is None.

orf commented on django/conf/global_settings.py:
```diff
@@ -642,3 +642,5 @@ def gettext_noop(s):
 SECURE_SSL_HOST = None
 SECURE_SSL_REDIRECT = False
+
+MODEL_DEFAULT_PK = 'django.db.models.AutoField'
```
That's an interesting point, but no I don't think so. It needs to be available/ready at the point of preparing the model, and I don't think information is ready at this point.

orf commented on django/contrib/admin/apps.py:
```diff
@@ -10,6 +10,7 @@ class SimpleAdminConfig(AppConfig):
     default_site = 'django.contrib.admin.sites.AdminSite'
     name = 'django.contrib.admin'
     verbose_name = _("Administration")
+    model_default_pk = 'django.db.models.AutoField'
```
For now, yes. We're in a bit of a pickle here because if we don't set this explicitly then there will be a large and annoying amount of migrations (or `AppConfig`'s) that need to be created by every user who wishes to use `BigAutoField`.

The migrations we currently ship create models with `AutoField`'s, so and we have no good story for changing that. I think a likely outcome of this MR is that third party apps that ship migrations (including `contrib` apps) will need to explicitly define their primary keys. Which could be spun as a good thing, as it forces authors to think about this rather than just going with `AutoField`.

charettes commented on django/core/checks/compatibility/django_3_2.py:
```diff
@@ -0,0 +1,19 @@
+
+W001 = Warning(
+    "The DEFAULT_MODEL_PK setting will default to 'django.db.models.BigAutoField "
+    "in Django 4.0. Set this setting to 'django.db.models.AutoField' or define "
+    "an explicit primary key to avoid extra migrations.",
```
Should this be part of `hint` instead? Also setting an explicit primary key on models won't actually address this warning as it will keep being emitted until `DEFAULT_MODEL_PK` is set.

I wonder if a per-model warning would be more appropriate when `settings.DEFAULT_MODEL_PK` is not defined , no explicit `primary_key` is specified, and the app config of the model doesn't define `default_model_pk`.

The warning could then hint at either defining the setting, adjusting the app config's `.default_model_pk` or explicitly declaring a `primary_key` on the model warned about.

orf replied:
I really like this idea, I've changed the PR to have per-model checks. I'm not 100% on the wording, but there are two cases here: 
1. Old projects with existing models
2. New projects created from templates that don't have a value for `DEFAULT_MODEL_PK`

So I've worded it to say set it to _either_ `BigAutoField` or `AutoField`, in case the previous wording encourages people to just blindly set it to `AutoField` on new projects.

pope1ni commented on django/db/models/fields/__init__.py:
```diff
@@ -2305,6 +2305,8 @@ class UUIDField(Field):
 
     def __init__(self, verbose_name=None, **kwargs):
         kwargs['max_length'] = 32
+        if kwargs.get("primary_key") and kwargs.get("auto_created"):
+            kwargs.setdefault("default", uuid.uuid4)
```
This feels to me like it is attempting to fudge a non-integer auto-field (`UUIDAutoField`) which has been discussed in the past. A lot more work is needed for this as there are lots of `isinstance(…, AutoField)` checks. See the discussions and implementation in #10680 that provide the initial cleanup preparing for generic auto-fields.

Should we be allowing any field type to be used? Or should we only be allowing `AutoFieldMixin` subclasses? Should the setting be `DEFAULT_AUTOFIELD` instead of `DEFAULT_MODEL_PK`?

I feel this PR should be limited to supporting specifying an alternative to the default `AutoField`. The brief IIRC was the desire to implement a mechanism such that we can default to `BigAutoField` without causing migration headaches for existing projects. Additional auto-created field types should be handled as a separate issue.

orf replied:
I like the idea of being able to specify any supported field as the primary key automatically. We support `UUIDField(primary_key=True)` right now (seemingly) without problems. I'll spend some time grepping around for the `isinstance` checks.

I'm not sure I agree that this should be handled as a different PR - you won't get an `AutoField` with `primary_key=True, auto_created=True` and without this change a `UUIDField` is pretty useless as a primary key field. To me they seem intricately linked.

pope1ni replied:
Hmm. Maybe I'm muddying the waters a little bit. Don't know.

Up till now the auto-generated field has always been an `AutoField`. In that sense it is tied to being an integer and having the value auto-generated by the database and returned via a `RETURNING` clause (or similar). I understand that you can manually override things in your model by specifying something like `UUIDField(primary_key=True)` - I have often done this too. But allowing it to take the place of `AutoField` as an auto-generated `id` field on the model is different. Hence my comments about checking for `isinstance(..., AutoField)` - there are a lot of hard-coded assumptions about the auto-generated field on a model.

I guess I'm thinking of something like `UUIDAutoField` where it is the database that actually provides the default value. We now have some support via `db_returning` for _creating_ new objects that return a value from the database. But we don't yet have the ability to specify something like `db_default` to choose a function to be executed on the database to generate that value, e.g. [`gen_random_uuid()`](https://www.postgresql.org/docs/13/functions-uuid.html) in PostgreSQL.

Maybe I'm making a mountain out of a molehill, but I just wanted to highlight this stuff for careful consideration. :male_detective: 

This change feels like a bit of a kludge so that we can get a sort-of "support auto-generated UUID primary key" :heavy_check_mark: in addition to the original concern of not using `BigAutoField` by default. (Not that I don't want it.)

orf replied:
I did a quick audit of `AutoField` usages within Django, and I can find two potential issues:

1. `bulk_create` does `fields = [f for f in fields if not isinstance(f, AutoField)]`
2. `forms.models.construct_instance` does `if not f.editable or isinstance(f, models.AutoField) or f.name not in cleaned_data: continue`

I'm not sure how to continue. On one hand, flexibility is nice and I can see a use for this in specific circumstances, but you are right that there might be additional issues we cannot see.

The issue with `UUIDAutoField` and using `db_default` in general is portability. As not all databases support generating a UUID (and PG needs an extension loaded) we'd surely need to fall back to `uuid.uuid4()` in some cases. I feel like this might require a lot of foundational work.

I think we should try to make this generic (I'll add test cases for all potential issues I can see), but we have to accept that there may be bugs that are discovered through this. In general we're pretty good at not assuming much about primary key fields as we do support different kinds, so I'd be really surprised if it does.

All that being said - this does increase the scope of the proposed change by a non-trivial amount. I'm obviously more keen about AutoField -> BigAutoField than anything else, so I'm happy to drop the generalization if that's the consensus.

pope1ni replied:
> I'm not sure how to continue.

Focus on making this work for switching `AutoField` to `BigAutoField` which was the primary brief. Other than careful decisions about naming the setting to avoid having to change it later, we should be able to consider how we are going to support non-auto fields for this as part of a separate PR later. That will include determining how to address the `isinstance(…, AutoField)` refactoring that was also punted from when I fixed the auto-field inheritance.

> As not all databases support generating a UUID (and PG needs an extension loaded) …

Most do, and loading an extension wouldn't be a deal breaker. We already have to do such a thing to enable `JSONField` for SQLite. At some point we need to accept that some extra effort is needed to enable certain functionality in certain backends. Also [`gen_random_uuid()`](https://www.postgresql.org/docs/13/functions-uuid.html) will be available without an extension in PostgreSQL 13 so gradually things will get easier.

orf replied:
> Focus on making this work for switching AutoField to BigAutoField which was the primary brief.

Ok, sounds good. For the implementation I am thinking that we keep the `DEFAULT_MODEL_PK` setting but raise an exception if it's not an `AutoField` subclass? I feel that it would be the most future proof approach rather than having some specific `USE_BIGINT_AUTOFIELD` setting.

I'll reduce the scope of the PR in the coming week.

pope1ni replied:
> I'll reduce the scope of the PR in the coming week.

Cool. Addressing this whole area in smaller, manageable chunks should hopefully result in ensuring we can get this landed.

> …than having some specific `USE_BIGINT_AUTOFIELD` setting.

Eww. That would be ghastly. 🤪 I can think of the following:

- `DEFAULT_MODEL_PK` - as currently implemented, but doesn't get across that it is a "field".
- `DEFAULT_AUTOFIELD` - my original suggestion, depends whether it is interpreted as an instance of `AutoField` or an "auto generated field". Perhaps `DEFAULT_AUTO_FIELD` instead…
- `DEFAULT_PK_FIELD` - could get confused with forms? Or maybe `DEFAULT_AUTO_PK_FIELD`…

charettes replied:
> Ok, sounds good. For the implementation I am thinking that we keep the DEFAULT_MODEL_PK setting but raise an exception if it's not an AutoField subclass? I feel that it would be the most future proof approach rather than having some specific USE_BIGINT_AUTOFIELD setting.

+1 to that.

orf replied:
I've changed the PR to raise a ValueError when the chosen PK field is not a subclass of `AutoField` 👍

felixxm commented on django/db/models/base.py:
```diff
@@ -1287,10 +1287,30 @@ def check(cls, **kwargs):
+                checks.Error(
+                    "Auto-created primary key used without defining a primary key type",
+                    hint=f"Set the DEFAULT_MODEL_PK setting or {app_config.__class__.__name__}.default_auto_field "
+                         f"attribute to either django.db.models.BigAutoField' or the legacy "
+                         f"'django.db.models.AutoField'.",
```
I wouldn't call `AutoField` a "legacy", we should also include `SmallAutoField`.

orf replied:
Is `SmallAutoField` ever the right choice for a primary key field? I really think not. The idea here is to guide people towards the right choice which is almost always `BigAutoField`. For completeness I've included it, but I think that it might be a mistake.

felixxm replied:
> Is SmallAutoField ever the right choice for a primary key field?

Why not, it can the right choice for any small dictionary tables if you really want to save space. I would even say that it's justified more often then `BigAutoField` :wink:

orf replied:
In one specific case it might be correct, and that would be: a small dictionary table, in an environment where saving between 2 and 4 bytes per row is important, and you are 100% sure you won't need to grow this table ever.

I think this is really specific enough to not warrant it appearing in a general error/warning message. Someone might look at this and say "I didn't know SmallAutoField exists, my data is small so it would be perfect!", which conflicts with the spirit of this work, which was to help people use a more future proof primary key field.

felixxm commented on django/db/models/base.py:
```diff
@@ -1287,10 +1287,30 @@ def check(cls, **kwargs):
+                         f"attribute to either django.db.models.BigAutoField' or the legacy "
+                         f"'django.db.models.AutoField'.",
+                    obj=cls,
+                    id="fields.E101"
```
This check should be documented in `docs/ref/checks.txt`.

felixxm commented on django/core/checks/__init__.py:
```diff
@@ -16,7 +16,6 @@
 import django.core.checks.translation  # NOQA isort:skip
 import django.core.checks.urls  # NOQA isort:skip
 
-
```
Revert unrelated change.

felixxm commented on docs/ref/applications.txt:
```diff
@@ -219,6 +219,14 @@ Configurable attributes
 
     By default, this attribute isn't set.
 
+.. attribute:: AppConfig.default_auto_field
```
`versionadded` is missing.

felixxm commented on docs/ref/settings.txt:
```diff
@@ -1336,6 +1336,16 @@ Default: ``''`` (Empty string)
+
+``DEFAULT_MODEL_PK``
+--------------------
+
```
`versionadded` is missing.

felixxm commented on tests/apps/tests.py:
```diff
@@ -101,8 +101,8 @@ def test_no_such_app_config(self):
         msg = (
-            "Module 'apps.apps' does not contain a 'NoSuchConfig' class. "
-            "Choices are: 'BadConfig', 'MyAdmin', 'MyAuth', 'NoSuchApp', "
+            "Module 'apps.apps' does not contain a class 'NoSuchConfig'. Choices are: "
+            "'BadConfig', 'ModelDefaultPkConfig', 'MyAdmin', 'MyAuth', 'NoSuchApp', "
```
Wrap at 79 chars.

felixxm commented on tests/apps/tests.py:
```diff
@@ -480,6 +480,23 @@ def test_multiple_paths_explicit_path(self):
+
+    @override_settings(INSTALLED_APPS=['apps.apps.PlainAppsConfig'])
+    def test_not_overridden_model_pk(self):
+        self.assertFalse(apps.get_app_config('apps')._is_model_default_pk_overridden)
```
`self.assertFalse()` -> `self.assertIs(..., False)`
`self.assertTrue()` -> `self.assertIs(..., True)`

felixxm commented on tests/runtests.py:
```diff
@@ -186,8 +186,9 @@ def no_available_apps(self):
     settings.SILENCED_SYSTEM_CHECKS = [
         'fields.W342',  # ForeignKey(unique=True) -> OneToOneField
         'fields.W903',  # NullBooleanField deprecated.
+        'fields.E101',  # BigAutoField migration
```
Do we need to silenced it globally? :thinking:

orf replied:
I think so - there are certainly a lot of apps inside the tests directory that don't set the PK setting. I'll see if I can remove this.

felixxm replied:
I pushed edits, it shouldn't be necessary anymore.

felixxm reviewed (commented):
@orf Thanks :+1: I left initial comments. I will try to review it in details in the next week.

felixxm commented:
buildbot

pope1ni commented on django/db/models/base.py:
```diff
@@ -1287,10 +1287,30 @@ def check(cls, **kwargs):
         return errors
 
+    @classmethod
+    def _check_default_pk(cls):
+        app_config = cls._meta.app_config
+        overridden_pk_appconfig = app_config._is_model_default_pk_overridden()
+        auto_created_pk = cls._meta.pk.auto_created
+        overridden_pk_setting = settings.is_overridden("DEFAULT_MODEL_PK")
+        if auto_created_pk and not overridden_pk_setting and not overridden_pk_appconfig:
+            return [
+                checks.Error(
+                    "Auto-created primary key used without defining a primary key type",
+                    hint=f"Set the DEFAULT_MODEL_PK setting or {app_config.__class__.__name__}.default_auto_field "
+                         f"attribute to either django.db.models.BigAutoField' or the legacy "
+                         f"'django.db.models.AutoField'.",
+                    obj=cls,
+                    id="fields.E101"
+                )
+            ]
+        return []
```
```suggestion
    def _check_default_auto_field(cls):
        if (
            cls._meta.pk.auto_created and
            not settings.is_overridden('DEFAULT_AUTO_FIELD') and
            not cls._meta.app_config._is_default_auto_field_overridden()
        ):
            return [
                checks.Error(
                    'Auto-created primary key used without defining a primary key type.',
                    hint=(
                        'Configure the DEFAULT_AUTO_FIELD setting or the %s.default_auto_field '
                        'attribute to point to a subclass of AutoField, e.g. '
                        "'django.db.models.BigAutoField'."
                    ) % cls._meta.app_config.__class__.__name__,
                    obj=cls,
                    id='fields.E101',
                ),
            ]
        return []
```

pope1ni commented on django/db/models/options.py:
```diff
@@ -250,7 +251,11 @@ def _prepare(self, model):
                 self.setup_pk(field)
             else:
-                auto = AutoField(verbose_name='ID', primary_key=True, auto_created=True)
+                pk_setting = getattr(self.app_config, "default_auto_field", settings.DEFAULT_MODEL_PK)
+                pk_class = import_string(pk_setting)
+                if not issubclass(pk_class, AutoField):
+                    raise ValueError(f"Primary key {pk_class} is not a subclass of AutoField")
+                auto = pk_class(verbose_name='ID', primary_key=True, auto_created=True)
```
```suggestion
                pk_setting = getattr(self.app_config, 'default_auto_field', settings.DEFAULT_AUTO_FIELD)
                pk_class = import_string(pk_setting)
                if not issubclass(pk_class, AutoField):
                    raise ValueError("Configured default auto field '%s' is not a subclass of AutoField." % pk_class)
                auto = pk_class(verbose_name='ID', primary_key=True, auto_created=True)
```

orf replied:
I've changed the quoting, but isn't this a good place to use f-strings? To me it's certainly more readable.

pope1ni commented on docs/ref/applications.txt:
```diff
@@ -219,6 +219,14 @@ Configurable attributes
+.. attribute:: AppConfig.default_auto_field
+
+    The implicit primary key type to add to models within this app. You can
+    use this to keep the historical :class:`~django.db.models.AutoField`
```
This isn't really "historical". We're just providing a mechanism for people to have a different default auto field class.

orf replied:
Wasn't the idea to migrate to using `BigAutoField` as the default in a future Django release? I've updated the wording though as it's currently not historical.

pope1ni commented on docs/ref/applications.txt:
```diff
@@ -219,6 +219,14 @@ Configurable attributes
+    use this to keep the historical :class:`~django.db.models.AutoField`
+    primary key type for third party applications.
+
+    By default this is the value of :setting:`DEFAULT_MODEL_PK`.
```
```suggestion
    By default this is the value of :setting:`DEFAULT_AUTO_FIELD`.
```

pope1ni commented on docs/ref/models/fields.txt:
```diff
@@ -415,10 +415,10 @@ cross-site scripting attack.
-don't need to set ``primary_key=True`` on any of your fields unless you want to
-override the default primary-key behavior. For more, see
-:ref:`automatic-primary-key-fields`.
+will automatically add a field with the type :setting:`DEFAULT_MODEL_PK` to
```
```suggestion
will automatically add a field with the type :setting:`DEFAULT_AUTO_FIELD` to
```

pope1ni commented on docs/ref/settings.txt:
```diff
@@ -1336,6 +1336,16 @@ Default: ``''`` (Empty string)
 Default tablespace to use for models that don't specify one, if the
 backend supports it (see :doc:`/topics/db/tablespaces`).
 
+.. setting:: DEFAULT_MODEL_PK
+
+``DEFAULT_MODEL_PK``
```
```suggestion
.. setting:: DEFAULT_AUTO_FIELD

``DEFAULT_AUTO_FIELD``
```

pope1ni commented on tests/admin_scripts/tests.py:
```diff
@@ -62,6 +62,7 @@ def write_settings(self, filename, apps=None, is_dir=False, sdict=None, extra=No
                 'DATABASES',
                 'ROOT_URLCONF',
                 'SECRET_KEY',
+                'DEFAULT_MODEL_PK',
```
```suggestion
                'DEFAULT_AUTO_FIELD',
```

pope1ni commented on tests/check_framework/test_model_checks.py:
```diff
@@ -358,3 +358,43 @@ class Meta:
+
+@isolate_apps('check_framework.apps.CheckAppConfigWithoutModelPK', attr_name='apps')
+@override_system_checks([checks.model_checks.check_all_models])
+class ModelDefaultPKTests(SimpleTestCase):
```
```suggestion
class ModelDefaultAutoFieldTests(SimpleTestCase):
```

pope1ni reviewed (commented):
I've had a quick blast through this with some thoughts.

I'm happy to pick this up if you don't have time for it at the moment?

orf commented on django/conf/global_settings.py:
```diff
@@ -649,3 +649,5 @@ def gettext_noop(s):
 SECURE_SSL_HOST = None
 SECURE_SSL_REDIRECT = False
+
+DEFAULT_MODEL_PK = 'django.db.models.AutoField'
```
Not sure what happened there! Fixed here and the other places 👍

orf commented on django/apps/config.py:
```diff
@@ -290,3 +290,14 @@ def ready(self):
+        from django.conf import settings
+        return settings.DEFAULT_MODEL_PK
+
+    def _is_model_default_pk_overridden(self):
```
Good idea - I've made it a property but kept the underscore prefix.

orf commented on django/apps/config.py:
```diff
@@ -290,3 +290,14 @@ def ready(self):
         Override this method in subclasses to run code when Django starts.
         """
+
+    @property
```
I think the only case where it will change at runtime is when running the Django tests. I'll see if making this a cached property has any impact with those.

orf commented:
Thanks for the review @pope1ni and @felixxm 🎉. I've made the changes you requested, but I did run into one corner case that requires a small change to the `Settings` object.

In `runtests.py` we set do the following to prevent a large number of `DEFAULT_AUTO_FIELD is not set` warnings:

```python
settings.DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'
```

However this does not cause `settings.is_overridden("DEFAULT_AUTO_FIELD")` to return True, despite it clearly being overridden (just not directly in the settings module itself). I've adjusted the `Settings` object to [track additions and deletions here](https://github.com/django/django/pull/13179/files#diff-b6f3108584c2db0b4c53253b6b3b5d5c60fbb3a82e7ed6544922fe689631fa43R225). Can you think of a better approach, and should I split this out into a separate PR?

felixxm reviewed (commented):
@orf Thanks for updates :rocket: We're close :1st_place_medal: 

Some adjustments to the ["Automatic primary key fields"](https://docs.djangoproject.com/en/3.1/topics/db/models/#automatic-primary-key-fields) paragraph are missing.

> However this does not cause settings.is_overridden("DEFAULT_AUTO_FIELD") to return True, despite it clearly being overridden (just not directly in the settings module itself). I've adjusted the Settings object to track additions and deletions here. Can you think of a better approach, and should I split this out into a separate PR?

Django doesn't [support ​altering settings at runtime](https://docs.djangoproject.com/en/dev/topics/settings/#altering-settings-at-runtime), and I don't think it should, see ticket-31165. IMO we should drop the last two commits, add `DEFAULT_AUTO_FIELD` to the [tests/test_sqlite.py](https://github.com/django/django/blob/master/tests/test_sqlite.py), and update CI configuration (I will do this).

felixxm commented:
@orf Do you have time to keep working on this? I would like to merge this in 2020 :calendar:  so I can push edits on my own if you're busy.

orf commented:
I'm a little bit busy this week, but I was planning to spend some time on it this weekend.

However please do go ahead and push any edits you'd like if you have time!

felixxm commented:
> Django doesn't [support ​altering settings at runtime](https://docs.djangoproject.com/en/dev/topics/settings/#altering-settings-at-runtime), and I don't think it should, see [ticket-31165](https://code.djangoproject.com/ticket/31165). IMO we should drop the last two commits, add `DEFAULT_AUTO_FIELD` to the [tests/test_sqlite.py](https://github.com/django/django/blob/master/tests/test_sqlite.py), and update CI configuration (I will do this).

I pushed small edits and moved handling not overridden `DEFAULT_AUTO_FIELD` to the tests. It's a bit hacky, but allows us to revert changes in `Settings` :rocket: I'm going to pushed more small edits ...

felixxm commented:
I will add `DEFAULT_AUTO_FIELD` to the CI configuration and rerun tests.

felixxm commented:
I added `_get_default_pk_class()` and improved tests for Model's options.

pope1ni commented on django/db/models/options.py:
```diff
@@ -217,6 +218,33 @@ def _format_names_with_class(self, cls, objs):
+            pk_class = import_string(pk_class_path)
+        except ImportError as e:
+            msg = f"The module '{pk_class_path}' could not be imported. "
+            print(self.app_config, hasattr(self.app_config, 'default_auto_field'))
```
```suggestion
```

felixxm commented:
buildbot, test on oracle.

felixxm commented on django/db/models/options.py:
```diff
@@ -250,7 +251,11 @@ def _prepare(self, model):
+                pk_setting = getattr(self.app_config, 'default_auto_field', settings.DEFAULT_AUTO_FIELD)
+                pk_class = import_string(pk_setting)
+                if not issubclass(pk_class, AutoField):
+                    raise ValueError(f"Primary key '{pk_class}' is not a subclass of AutoField")
```
After rethinking, a new system check is not necessary. Let's leave it here.

felixxm commented:
I fixed crash when `DEFAULT_AUTO_FIELD` is None, added tests and improved error messages.

orf commented on django/db/models/options.py:
```diff
@@ -218,6 +218,32 @@ def _format_names_with_class(self, cls, objs):
+            raise ImproperlyConfigured(msg) from e
+        if not issubclass(pk_class, AutoField):
+            raise ValueError(
+                f"Primary key '{pk_class_path}' must subclass AutoField."
```
Do you think it would be helpful to include the app config path in this message, similar to how you do it above on line 234?

felixxm replied:
Done.

felixxm commented on django/db/models/base.py:
```diff
@@ -1290,10 +1290,32 @@ def check(cls, **kwargs):
+            not cls._meta.app_config._is_default_auto_field_overridden
+        ):
+            return [
+                checks.Error(
+                    'Auto-created primary key used without defining a primary key type.',
+                    hint=(
+                        'Configure the DEFAULT_AUTO_FIELD setting or the %s.default_auto_field '
+                        'attribute to point to a subclass of AutoField, e.g. '
+                        "'django.db.models.BigAutoField'."
+                    ) % cls._meta.app_config.__class__.__name__,
+                    obj=cls,
+                    id='fields.E101',
```
IMO we shouldn't raise an `Error` here, `Warning` sound fine. We don't really want to force all users to `DEFAULT_AUTO_FIELD` to their settings. Moreover we will probably remove this check after changing the default. I will push edits.

felixxm commented:
@orf I pushed all edits (except release notes). Do you want to take a look?

carltongibson reviewed (commented):
Hi all. Thanks for the work here. Looks good. 

I've pushed some suggested edits for the release notes. Please the see the `[...ADD REASON...]` block, where I could do with a suggestion 😄 

Just a double-check: as well as adding the ability, +1 super, are we really sure about changing the default? It strikes me that a _Very small proportion indeed™_ of projects are ever going to need this. No matter the warning, or the lead-time, there are going to be a lot of unneeded/unwanted/quite-disruptive migrations created and run because of that change. Is this not a _Pro™_ feature that should be opt-in? (Don't shoot me! 🙂)

pope1ni commented:
> Just a double-check: as well as adding the ability, +1 super, are we really sure about changing the default? It strikes me that a Very small proportion indeed™ of projects are ever going to need this. No matter the warning, or the lead-time, there are going to be a lot of unneeded/unwanted/quite-disruptive migrations created and run because of that change. Is this not a Pro™ feature that should be opt-in? (Don't shoot me! 🙂)

We've only set the default in the `startproject` template to `BigAutoField`. This ensures that new projects won't suffer from becoming larger projects that painfully need the column migrating years down the line. I've been there.

In my mind we shouldn't need to change the value in the default global settings away from `AutoField`. That will avoid foisting new and difficult migrations on developers. `DEFAULT_AUTO_FIELD` can be explicitly changed in an existing project if someone wants or needs to go through the pain of updating everything. Equally, I potentially see this as the way for solving ticket-56 with a `Positive(Big)AutoField`, and also maybe allowing a `UUIDAutoField`. This setting allows these to be set by default for a project.

But there was the challenge of third-party apps. To solve this `AppConfig.default_auto_field` was added. It defaults to the value of `DEFAULT_AUTO_FIELD`. This allows for apps to behave differently to the project as a whole, thus preventing unnecessary breakage. Note that this has been set to `AutoField` for the apps that are part of Django for this reason. It also allows for the apps in a project to be updated independently for a phased change in the default auto field, or for only those apps that someone wants to change.

I guess the only point I'm not clear on is what happens if someone wanted to use `BigAutoField` with `django.contrib.auth`, say, because then they'd need to generate migrations? (Subclassing the `AppConfig` and changing `default_auto_field` should allow for that, but then they'd conflict with future changes to the app.) The documentation doesn't cover this - I think you'd be forced into forking the app?

But, yes, I'm not sure we ever need to change the global default as this is handled by `startproject` for new projects. Everyone else will only need to change as required.

orf commented:
> Just a double-check: as well as adding the ability, +1 super, are we really sure about changing the default? It strikes me that a _Very small proportion indeed™_ of projects are ever going to need this

I like sensible defaults and I think `BigAutoField` is _the_ sensible, most general default for primary keys. There are exceptions, but the current default essentially implies "your table will never grow past this very reachable limit" which IMO isn't a general, sensible default.

If there was a way to change the default for new models whilst keeping backwards compatibility with existing models then  I would push for that, but there isn't a way I can think of to do that 😭. I really wish there was. But for now a "setting that we hope is correctly added when you start a project" is a good compromise I think.

> I guess the only point I'm not clear on is what happens if someone wanted to use BigAutoField with django.contrib.auth, say, because then they'd need to generate migrations? (Subclassing the AppConfig and changing default_auto_field should allow for that, but then they'd conflict with future changes to the app.) The documentation doesn't cover this - I think you'd be forced into forking the app?

This is an inherent problem with apps shipping migrations if the migrations now can change based on outside state. I can't think of a way around this, so the `default_auto_field` is more to prevent extra migrations from being generated rather than to allow someone to customize `default_auto_field` within an app like `contrib.auth` that ships migrations.

orf commented on docs/topics/db/models.txt:
```diff
@@ -259,12 +259,13 @@ details can be found in the :ref:`common model field option reference
+<django.apps.AppConfig.default_auto_field>` or globally in the
+:setting:`DEFAULT_AUTO_FIELD` setting. For example::
 
     id = models.AutoField(primary_key=True)
```
Maybe we could change this to now show `BigAutoField`, in the theme of "encouraging people to use BigAutofield's"?

```suggestion
    id = models.BigAutoField(primary_key=True)
```

orf commented:
> @orf I pushed all edits (except release notes). Do you want to take a look?

Thank you for pushing this over the line @felixxm! 🎉. The changes look great, I'm still not sure on the appropriate wording for the docs, but I'll have a think.

orf commented on docs/releases/3.2.txt:
```diff
@@ -53,6 +53,46 @@ needed. As a consequence, it's deprecated.
+
+When defining a model, if no field in a model is defined with
+:attr:`primary_key=True <django.db.models.Field.primary_key>` an implicit
+primary key will be added. In order to better [...ADD REASON...], the type of
```
> In order to change this default without requiring large numbers of changes to your models,

maybe?

carltongibson commented:
OK, thanks both. 

> In my mind we shouldn't need to change the value in the default global settings away from AutoField. That will avoid foisting new and difficult migrations on developers. 

We should decide the plan here. The release notes as are say that we will change the default. Also, the system check seems unnecessary if it’s not going to change.

pope1ni commented:
Ooh. Another thought: We should probably make `startapp` set `AppConfig.default_auto_field` to `BigAutoField` too. That will help for models in new apps. It won't help for new models in existing apps, but it's something.

felixxm commented:
> Ooh. Another thought: We should probably make `startapp` set `AppConfig.default_auto_field` to `BigAutoField` too. That will help for models in new apps. It won't help for new models in existing apps, but it's something.

Done. I also squashed my previous edits.

felixxm commented:
> OK, thanks both.
> 
> > In my mind we shouldn't need to change the value in the default global settings away from AutoField. That will avoid foisting new and difficult migrations on developers.
> 
> We should decide the plan here. The release notes as are say that we will change the default. Also, the system check seems unnecessary if it’s not going to change.

I'm also not sure about changing the value in the global settings. It seems unnecessary, and removing the system check would significantly reduce the complexity of this patch (CI changes, monkey patching, etc.).

orf commented:
I strongly, strongly believe that bigint is _the_ sensible default for any modern web application. Yes, it's not always needed, but in the cases where it's not needed the difference between bigint and int is negligible. Where it _is_ needed the difference is **huge**. I won't rehash the arguments I've given previously, but I would push for the following:
1. We merge this with the system check
2. Users are encouraged to set the required setting in their app settings file
3. At a later date we switch the default to bigint

It's great to add the setting to `startproject` and `startapp`, but that doesn't cover all bases - especially with projects that are created by templates other than the default Django ones. By adding a system check we make people add an explicit opt-in to the old `int` (or the new `bigint`) configuration, and in the future we flip the default. Users who haven't added an explicit setting can just add one while upgrading, whilst all new projects are ensured to use a more sensible default.

felixxm commented:
@orf Agreed, I squashed commits and adjusted release notes.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on selenium.

carltongibson commented:
Yep, fine. Thanks!

pope1ni commented:
🚀 Let's do this!

felixxm commented:
Thanks y'all :1st_place_medal:  :rocket:
